rfcs/records.md

```diff
@@ -0,0 +1,201 @@
+local r = new X(1, 2)
+```
+
+### Generic records
+
+> TODO: This needs some thought; e.g., do we support explicit specification of record arguments at construction time and what's the syntax for that? Do we need to do this in the first proposal? This likely makes packed records effectively impossible to support at runtime without a huge amount of work, does that matter?
+
+### Type checking records
+
+Records defined via a `record` statement can be used in type annotations as usual. The unification rules say that a table can unify with a record if the fields
+match, which makes records similar to sealed tables from the type checking perspective. (note, this is hand wavy on subtyping rules)
+
+> TODO: How do you export a record type? `export record` would be straightforward but potentially conflicts with future export statements for functions/values.
+> Alternatively, is `export type Record = Record` too awkward?
+
+The methods defined on the record object are type checked as usual, with one exception - the implicit `self` has the type of the record. This is crucial because
+this is the one big issue we can't resolve with metatable-based OOP for tables, but it works for records because `:` is slightly more magical.
+
+When `self` is explicit, the type needs to be specified manually, e.g. these definitions are equivalent:
+
+```
+function Point:sum(): number
+    return self.x + self.y
+end
+
+function Point.sum(self: Point): number
+    return self.x + self.y
+end
+```
+
+Of course, the type checker also knows that the record type has the metatable with the inferred type of the record shape.
+
+> TODO: Does the type checker need to understand the internal structure of the shape so that type checking works across modules, or is simply modeling this
+> as a metatable sufficient?
+
+### Object modeling
+
+> TODO: records don't support implementation inheritance and why it's a good thing
+
+## Drawbacks
+
+Adding a new data type that is cross-cutting (across syntax, semantics/compiler, semantics/type checking, and runtime) results in added complexity.
+
+The rigidity of records may make some applications hesitate to adopt them; e.g. you can't simply add a new field at a random point in the program, which some would
+argue makes the language less dynamic and therefore less convenient.
+
+Not enforcing type compatibility for typed records at runtime may make it difficult for us to optimize record storage more by removing the type tags (which could
+make record objects ~2x more efficient in some cases).
+
+## Alternatives
+
+Instead of using explicit record types, we can make the VM recognize shapes of objects automatically, just like JavaScript implementations do. This requires a
+substantial amount of complicated machinery and heuristics, and likely can't be as efficient as records in the long run, but it can result in close efficiency
+without any changes to existing programs. This, however, doesn't make type safety any easier.
```

Comment:
This seems a bit strong. All the shape machinery, speculative optimization, etc is here, the differences are a) making shapes immutable, and b) the speculation is based on ahead-of-time analysis not tracing. I
'm not sure that adding nominal types makes the type system easier.

Reply:
This is referring specifically to the issues around table-based type checking we have not resolved today. The proposed structure provides an obvious solution to tying together the shape of the table and the methods on a type level plus an obvious syntax for specifying field types, something we can't do with metatables today.

Reply:
On the former, the proposal here is much simpler from the runtime perspective than automatic shape optimization.

Reply:
Well yes, fixing tables is just as vaporware as records. I don't see why the runtime is much simpler though, if we added immutable shapes to tables then the code for lookups looks very similar to that for records, but with a fallthrough case of looking up a hash table rather than erroring.

The thing that is a bit more complex is getting the shape information out of the type system, since we'd probably want to cache it. And any code which didn't get typechecked (e.g. anything inside an eval) is going to get no shape optimization.

Reply:
> And any code which didn't get typechecked (e.g. anything inside an eval) is going to get no shape optimization.

And that's the crux. This proposal doesn't rely on the type system for efficiency - it provides efficiency regardless of whether we ran type checking. Today it's not practical to run type checking during compilation because we don't provide termination or timeliness guarantees. Eventually we'll hopefully get to the place where it is, but even there I think it's very important to maintain unsurprising performance, and we're likely going to have many cases where even when type checking is active, we don't actually infer record types because they are not annotated (e.g. assignment/lookup).

Additionally, to perform the optimization you're talking about we need cross-module compilation. This fundamentally alters the current compilation model in a way that is tremendously difficult to deploy due to dynamic nature of require in our current system.

Shape optimization is fundamentally prone to deoptimization which can result in performance cliffs, results in having to maintain complex caches of this information and has problems around object reallocation. It's undoubtedly possible to implement, as proven by various JS VMs. But the core benefit of Luau is that we own the entire stack and we can make decisions that result in optimal outcomes across the entire stack, which minimizes the complexity and maximizes predictability of performance and behavior.